bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections/ConcurrentLongLongPairHashMap.java

```diff
@@ -376,7 +380,16 @@ private boolean remove(long key1, long key2, long value1, long value2, int keyHa
                 }
 
             } finally {
-                unlockWrite(stamp);
+                if (size < resizeThresholdBelow) {
```

Comment:
Why are we ohtting this code in the 'finally' block?

Reply:
> Why are we ohtting this code in the 'finally' block?

to shrink when we go below a certain threshold. @eolivelli

Reply:
> Also we should add test cases that verify that the map is behaving as coded

Added test, PTAL,thanks! @eolivelli

Reply:
We out in the finally block onlybl code that must be always executed.
I would move this part inside the try block

Reply:
I think this should be executed every time (similar to the expansion logic): after each remove, it should be judged whether to shrink or not：
https://github.com/apache/bookkeeper/blob/7018d2fff35de96a90a1008366e38580e448fb84/bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections/ConcurrentLongLongPairHashMap.java#L293-L320

Reply:
Why are we comparing here 'size' and we are using 'usedBuckets' while comparing for the other threshold?

Reply:
Because when get item, EmptyKey is used as the end mark, if you use size to judge when expanding, it may cause EmptyKey not found in map, which leads to infinite loop when get item.

Reply:
I agree to add a new configuration autoShrink to control whether to shrink or not. @eolivelli 

Then we have to solve a problem: when autoShrink=false, even if the map has a lot of space, the expansion will be triggered frequently, because the condition of our expansion is not to see the actual use size of the map.

Reply:
> Why are we comparing here 'size' and we are using 'usedBuckets' while comparing for the other threshold?

When shrinking, size is used because size is the size actually occupied by the map, so as to determine whether the current wasted memory ratio is particularly large. If there is too much, then we need to reclaim some memory.

Reply:
I agree that the condition foe the expansion is not good and probably we should fix it, but maybe not in this PR